extensions/datasource/deployment/src/main/java/io/quarkus/datasource/deployment/devservices/DevServicesDatasourceProcessor.java

```diff
@@ -172,14 +189,47 @@ public void run() {
             closeBuildItem.addCloseTask(closeTask, true);
         }
         databases = runningDevServices;
-        cachedProperties = propertiesMap;
+        cachedProperties = getRestartTriggeringPropertiesValues();
         for (RunningDevService database : databases) {
             devServicesResultBuildItemBuildProducer.produce(database.toBuildItem());
         }
         return new DevServicesDatasourceResultBuildItem(results);
     }
 
-    private String trim(String optional) {
+    /**
+     * Returns a map of properties that are can trigger a datasource dev service restart if modified.
+     * It excludes EXCLUDED_PROPERTIES and includes only RESTART_TRIGGERING_PROPERTIES.
+     * If a property ends with {@code .} then it will also trigger a restart on any sub properties change
+     */
+    private static Map<String, String> getRestartTriggeringPropertiesValues() {
+        Map<String, String> datasourceProperties;
+        try {
+            datasourceProperties = ConfigProvider.getConfig()
+                    .unwrap(SmallRyeConfig.class)
+                    .getOptionalValues("quarkus.datasource",
+                            String.class,
+                            String.class)
+                    .orElse(Map.of());
+        } catch (NoSuchElementException e) {
```

Comment:
This happens in integration tests _integration-tests/jpa-postgresql_ the maven property `postgresq.url` is not resolved unknown at the time the datasource properties are queried, and the test fails.

I don't know if this is a normal an expected behaviour, so i put this catch. We can remove it if itdoesn't make sense

Reply:
This is concerning... and no, I don't think catching and ignoring the exception is acceptable here.

@radcortez do you think this could be related to the class `io.quarkus.it.jpa.postgresql.OverrideJdbcUrlBuildTimeConfigSource` you added to that module?

Reply:
Yes it is.

For now, I've just pushed an update to disable expression expansion in that piece of code, but this is a bug in SmallRye Config. Any call to `getOptional*` should not give an error if the expression cannot be expanded.

This also means that we cannot reliably monitor configuration names that target an expression, at least with the current implementation. Because an expression may only be available at runtime, the cache would have to be built in a runtime recorder.

Reply:
It seems your fix allowed one test to pass, but uncovered other errors...

> This also means that we cannot reliably monitor configuration names that target an expression, at least with the current implementation. Because an expression may only be available at runtime, the cache would have to be built in a runtime recorder.

This code starts dev services and needs to happen in build steps, though.
Are you saying we cannot use `getOptional*` in this PR and must find another way?

Reply:
Is the solution, with the catch even if not covering all cases, is it not already better than the current behaviour of Always restarting ?
Because in the worst case of a thrown error, it will simply mean that the dev services will be restarted each time. i could not reproduce it ourside of the test by providing runtime peoperties, and variable substitution.

Alternatively, the solution on the first commit on the PR works without those issues, although with a clunky solution.

Reply:
> Is the solution, with the catch even if not covering all cases, is it not already better than the current behaviour of Always restarting ?

It is not.

The current solution does not "always restart" incorrectly. It restarts incorrectly in a specific case you reported, when changing quotes.

As far as I know, the current solution does not restart "all the time" when a property value contains `${postgres.url}`. The new solution would.

I'm open to reverting to your original solution before your switched to `getOptionalValues()`, but my comment would still stand: 
https://github.com/quarkusio/quarkus/pull/41736#discussion_r1668653666

A third solution would be to extract the data from `DataSourceBuildTimeConfig` directly. But that solution is obvious, so much simpler, that I'm pretty sure there was a good reason not to do that.

In any case... let's wait for @radcortez 's insights before doing anything else.

Reply:
> The current solution does not "always restart" incorrectly. It restarts incorrectly in a specific case you reported, when changing quotes.
> 
Just to be precise it Always restarts, even with no changes, if there is at least one named datasource, quoted or not.

Reply:
> This code starts dev services and needs to happen in build steps, though.
> Are you saying we cannot use `getOptional*` in this PR and must find another way?

You can. The expression expansion issue that I brought up also happened with the old implementation, but I guess it is a niche case.

To summarize, we cannot safely expand runtime expressions during build time (where we populate the cache), because the configuration source may not be available at this point. So if we cache `quarkus.datasource.jdbc.url` to `${postgres.url}`, the user will change the value of `postgres.url`, which would not trigger the restart. I guess we could add such expressions to the monitor list.

Reply:
> A third solution would be to extract the data from `DataSourceBuildTimeConfig` directly. But that solution is obvious, so much simpler, that I'm pretty sure there was a good reason not to do that.

Probably because mappings didn't support `equals` and `hashcode` which they do now. So you could either do it for the entire object, or selected pieces.

Reply:
> Probably because mappings didn't support `equals` and `hashcode` which they do now. So you could either do it for the entire object, or selected pieces.

So we can simply call `equals` between two `DataSourceBuildTimeConfig` ?

Or do we still need a map with something like that we will compare ? :

```java
Map<String, Object> res = new HashMap<>();
for (var datasource : dataSourcesBuildTimeConfig.dataSources().entrySet()) {
    String name = datasource.getKey();
    DataSourceBuildTimeConfig config = datasource.getValue();
    res.put(name + ".db-kind", config.dbKind());
    res.put(name + ".db-version", config.dbVersion());
    res.put(name + ".devservices.command", config.devservices().command());
    res.put(name + ".devservices.container-env", config.devservices().containerEnv());
    res.put(name + ".devservices.container-properties.", config.devservices().containerProperties());
    res.put(name + ".devservices.db-name", config.devservices().dbName());
    res.put(name + ".devservices.enabled", config.devservices().enabled());
    res.put(name + ".devservices.image-name", config.devservices().imageName());
    res.put(name + ".devservices.init-script-path", config.devservices().initScriptPath());
    res.put(name + ".devservices.password", config.devservices().password());
    res.put(name + ".devservices.port", config.devservices().port());
    res.put(name + ".devservices.properties", config.devservices().properties());
    res.put(name + ".devservices.reuse", config.devservices().reuse());
    res.put(name + ".devservices.username", config.devservices().username());
    res.put(name + ".devservices.volumes", config.devservices().volumes());
}
```

Reply:
> So we can simply call equals between two DataSourceBuildTimeConfig ?

No, because then you'd be restarting too often, e.g. when `DataSourceBuildTimeConfig#healthExclude` changes.

> Or do we still need a map with something like that we will compare ? :

That's an option... or you hold the information you are interested in, and only that information, in something like this:

```
private record DevServiceConfigCache(
    Optional<String> dbKind,
    Optional<String> dbVersion,
    DevServicesBuildTimeConfig devservices
)
```

Reply:
Actually scratch that, since apparently we don't want to restart when `quarkus.datasource.devservices.enabled` changes... your solution with a Map is better, more flexible. And more backportable.